# Lab Inferential Statistics

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import ttest_1samp
from scipy.stats import t

## One tailed T-test

In a packing plant, a machine packs cartons with jars. It is supposed that a new machine will pack faster on the average than the machine currently used. To test that hypothesis, the times it takes each machine to pack ten cartons are recorded. The results, in seconds, are shown in the tables in the file *files_for_lab/machine.txt*. Assume that there is sufficient evidence to conduct the t test, does the data provide sufficient evidence to show if one machine is better than the other?

In [2]:
machine = pd.read_csv('files_for_lab/machine_new.txt', sep='\t')
machine

New machine      Old machine
0         42.1             42.7
1         41.0             43.6
2         41.3             43.8
3         41.8             43.3
4         42.4             42.5
5         42.8             43.5
6         43.2             43.1
7         42.3             41.7
8         41.8             44.0
9         42.7             44.1

In [3]:
machine.columns = ['New machine', 'Old machine']

* H0: mean time of new machine <= mean time of old machine -> mu(43.230000000000004) >= x(42.14)
* H1: mean time of new machine > mean time of old machine -> mu(43.230000000000004) < x(42.14)
* Sample = 10
* One sided (rejection area on the right side)
* Confidence interval = 0.95

In [4]:
print(machine['Old machine'].mean())
print(machine['New machine'].mean())

43.230000000000004
42.14


In [5]:
stat, pval = ttest_1samp(machine['New machine'], popmean=43.230000000000004, alternative='greater') #lower??

In [6]:
print(pval)
print(stat)

0.999651681196162
-5.04331853503833


In [7]:
#ppf = critical value vs stat-> If 
t.ppf(0.95, df=9)

1.8331129326536333

In [8]:
#cdf=p_value vs alpha
st.t.cdf(stat, df=9)

0.00034831880383795163

In [9]:
ttest_1samp(machine['New machine'], 43.230000000000004,  alternative='greater') #why a I getting different results?

TtestResult(statistic=-5.04331853503833, pvalue=0.999651681196162, df=9)

We have to accept H0 because p-value=0.999651681196162 is > alpha = 0.05 and stat = -5.043 < critical value = 1.83

## Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon (file files_for_lab/pokemon.csv). Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. Compare the two columns to see if there is a statistically significant difference between them and comment your result.

In [10]:
pokemon = pd.read_csv('files_for_lab/pokemon.csv')
pokemon

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

* Ho: attack = defense
* H1: attack != defense
* Sample = 30
* Confidence interval = 0.95
* Two sided

In [11]:
sample = pokemon.sample(30, random_state=16)
sample['difference'] = sample['Attack']-sample['Defense']
sample.head(30)

#                   Name    Type 1    Type 2  Total   HP  Attack  \
505  455              Carnivine     Grass       NaN    454   74     100   
324  300                 Skitty    Normal       NaN    260   50      45   
510  460              Abomasnow     Grass       Ice    494   90      92   
598  538                  Throh  Fighting       NaN    465  120     100   
552  493                 Arceus    Normal       NaN    720  120     120   
551  492       ShayminSky Forme     Grass    Flying    600  100     103   
208  193                  Yanma       Bug    Flying    390   65      65   
574  515                Panpour     Water       NaN    316   50      53   
528  476              Probopass      Rock     Steel    525   60      55   
741  673                 Gogoat     Grass       NaN    531  123     100   
351  321                Wailord     Water       NaN    500  170      90   
327  302    SableyeMega Sableye      Dark     Ghost    480   50      85   
307  283                Surskit       Bug     Water    269   40      30   
20    16                 Pidgey    Normal    Flying    251   40      45   
547  489                 Phione     Water       NaN    480   80      80   
787  711    GourgeistSuper Size     Ghost     Grass    494   85     100   
679  618               Stunfisk    Ground  Electric    471  109      66   
216  201                  Unown   Psychic       NaN    336   48      72   
525  474              Porygon-Z    Normal       NaN    535   85      80   
627  566                 Archen      Rock    Flying    401   55     112   
620  559                Scraggy      Dark  Fighting    348   50      75   
780  710  PumpkabooAverage Size     Ghost     Grass    335   49      66   
662  601              Klinklang     Steel       NaN    520   60     100   
422  382    KyogrePrimal Kyogre     Water       NaN    770  100     150   
42    37                 Vulpix      Fire       NaN    299   38      41   
416  379              Registeel     Steel       NaN    580   80      75   
383  351               Castform    Normal       NaN    420   70      70   
498  448    LucarioMega Lucario  Fighting     Steel    625   70     145   
592  532                Timburr  Fighting       NaN    305   75      80   
728  660              Diggersby    Normal    Ground    423   85      56   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  difference  
505       72       90       72     46           4      False          28  
324       45       35       35     50           3      False           0  
510       75       92       85     60           4      False          17  
598       85       30       85     45           5      False          15  
552      120      120      120    120           4       True           0  
551       75      120       75    127           4       True          28  
208       45       75       45     95           2      False          20  
574       48       53       48     64           5      False           5  
528      145       75      150     40           4      False         -90  
741       62       97       81     68           6      False          38  
351       45       90       45     60           3      False          45  
327      125       85      115     20           3      False         -40  
307       32       50       52     65           3      False          -2  
20        40       35       35     56           1      False           5  
547       80       80       80     80           4      False           0  
787      122       58       75     54           6      False         -22  
679       84       81       99     32           5      False         -18  
216       48       72       48     48           2      False          24  
525       70      135       75     90           4      False          10  
627       45       74       45     70           5      False          67  
620       70       35       70     48           5      False           5  
780       70       44       55 

In [12]:
sample_diff_mean, sample_diff_std = sample['difference'].mean(), sample['difference'].std(ddof=1)
sample_diff_mean, sample_diff_std

(5.433333333333334, 34.54101343613707)

In [13]:
t = sample_diff_mean / ( sample_diff_std / np.sqrt(sample.shape[0]) )
print("The mean of our samples differences is: {:.2f}".format(sample_diff_mean))
print("The standard deviation of our samples differences is: {:.2f}".format(sample_diff_std))
print("Our t statistics is: {:.2f}".format(t))

The mean of our samples differences is: 5.43
The standard deviation of our samples differences is: 34.54
Our t statistics is: 0.86


In [14]:
df= sample.shape[0] - 1
df

29

We have to use T-test because it's < 30

In [15]:
tc_upper = st.t.ppf(1-(0.05/2),df= sample.shape[0] - 1)
tc_lower = st.t.ppf(1-(0.95/2),df= sample.shape[0] - 1)
print(tc_upper)
print(tc_lower)

2.045229642132703
0.06325175732602374


In [16]:
st.t.cdf(t,df = sample.shape[0] - 1)

0.8020044806100265

In [17]:
st.ttest_rel(sample['Attack'],sample['Defense'])

TtestResult(statistic=0.8615726445356495, pvalue=0.39599103877994724, df=29)

**Conclusion:** Our statistic(0.86) is inside the acceptance area [0.063, 2.045] and the p-value(0.40) is bigger than alpha(0.05) therefore we have to **accept the H0**.

## ANOVA

### Part 1

In this activity, we will look at another example. Your task is to understand the problem and write down all the steps to set up ANOVA. After the next lesson, we will ask you to solve this problem using Python. Here are the steps that you would need to work on: - Null hypothesis - Alternate hypothesis - Level of significance - Test statistic - P-value - F table

**Setting ANOVA:**
* H0: all the group means (the power of each microprocessor) are equal
* H1: any of the means is not equal
* Level of significance: 95%
* Test statistic:
* P-value
* F table

Suppose you are working as an analyst in a microprocessor chip manufacturing plant. You have been given the task of analyzing a plasma etching process with respect to changing Power (in Watts) of the plasma beam. Data was collected and provided to you to conduct statistical analysis and check if changing the power of the plasma beam has any effect on the etching rate by the machine. You will conduct ANOVA and check if there is any difference in the mean etching rate for different levels of power. You can find the data *anova_lab_data.xlsx* file in the files_for_lab folder

* State the null hypothesis
* State the alternate hypothesis
* What is the significance level
* What are the degrees of freedom of model, error terms, and total DoF
Data was collected randomly and provided to you in the table as shown: link to the image - Data

In [18]:
data = pd.read_excel('files_for_lab/anova_lab_data.xlsx')
data

Power   Etching Rate
0   160 W          5.43
1   180 W          6.24
2   200 W          8.79
3   160 W          5.71
4   180 W          6.71
5   200 W          9.20
6   160 W          6.22
7   180 W          5.98
8   200 W          7.90
9   160 W          6.01
10  180 W          5.66
11  200 W          8.15
12  160 W          5.59
13  180 W          6.60
14  200 W          7.55

In [19]:
data.columns=['Power','Etching Rate']

In [20]:
data['Power'].value_counts()

160 W    5
180 W    5
200 W    5
Name: Power, dtype: int64

In [21]:
group_df = data.groupby('Power')['Etching Rate'].agg(power_mean='mean',Samples='size').reset_index()
group_df

Power  power_mean  Samples
0  160 W       5.792        5
1  180 W       6.238        5
2  200 W       8.318        5

In [22]:
data['power_count'] = data.groupby('Power').cumcount()
data_pivot = data.pivot(index='power_count', columns = 'Power', values = 'Etching Rate')
data_pivot.columns = ['Power_'+str(x) for x in data_pivot.columns.values]
data_pivot.head()

Power_160 W  Power_180 W  Power_200 W
power_count                                       
0                   5.43         6.24         8.79
1                   5.71         6.71         9.20
2                   6.22         5.98         7.90
3                   6.01         5.66         8.15
4                   5.59         6.60         7.55

In [23]:
data_pivot.mean()

Power_160 W    5.792
Power_180 W    6.238
Power_200 W    8.318
dtype: float64

### Part 2

* In this section, use the Python to conduct ANOVA.
* What conclusions can you draw from the experiment and why?

In [24]:
st.f_oneway(data[data['Power'] == '160 W']['Etching Rate'], data[data['Power'] == '180 W']['Etching Rate'], 
            data[data['Power'] == '200 W']['Etching Rate'])[1]

7.506584272358903e-06

In [25]:
st.f_oneway(data_pivot['Power_160 W'], data_pivot['Power_180 W'], data_pivot['Power_200 W'])

F_onewayResult(statistic=36.87895470100505, pvalue=7.506584272358903e-06)

We have to reject Ho because p-value < alpha.